File: simulation/swarm/agent.cc
#include "simulation/swarm/agent.h"

#include <Eigen/Dense>
#include <boost/numeric/odeint.hpp>
#include <boost/numeric/odeint/external/eigen/eigen.hpp>
#include <cmath>
#include <cstdbool>
#include <stdexcept>

#include "absl/strings/str_format.h"
#include "simulation/swarm/proto/agent.pb.h"
#include "simulation/swarm/proto/static_config.pb.h"
#include "simulation/swarm/proto/transformation.pb.h"
#include "simulation/swarm/utils/constants.h"

namespace swarm::agent {

Agent::Agent(const AgentConfig& config, const double t_creation,
             const bool ready)
    : Agent(config.initial_state(), t_creation, ready) {
  dynamic_config_ = config.dynamic_config();
  plotting_config_ = config.plotting_config();
  submunitions_config_ = config.submunitions_config();
}

Agent::Agent(State initial_state, const double t_creation, const bool ready)
    : t_creation_(t_creation),
      state_(std::move(initial_state)),
      flight_phase_(ready ? FlightPhase::READY : FlightPhase::INITIALIZED) {
  // Add the initial state to the history.
  state_history_.Add(state::StateHistory::Record(t_creation_, hit_, state_));
}

void Agent::AssignTarget(Agent* target) {
  target_ = target;
  target_model_ = std::make_unique<ModelAgent>(target->state());
}

void Agent::CheckTarget() {
  if (has_assigned_target() && target_->hit()) {
    UnassignTarget();
  }
}

void Agent::UnassignTarget() {
  target_ = nullptr;
  target_model_.release();
}

void Agent::MarkAsHit() {
  hit_ = true;
  // Update the latest hit boolean in the history.
  state_history_.back().hit = true;
  flight_phase_ = FlightPhase::TERMINATED;
}

bool Agent::HasHitTarget() const {
  if (!has_assigned_target()) {
    return false;
  }

  // Calculate the distance to the target.
  const auto position = GetPosition();
  const auto target_position = target_->GetPosition();
  const auto distance = (target_position - position).norm();

  // A hit is recorded if the target is within the agent's hit radius.
  const auto hit_radius = static_config().hit_config().hit_radius();
  return distance <= hit_radius;
}

void Agent::SetState(const State& state) {
  state_ = state;
  state_history_.back().state = state;
}

Agent::PrincipalAxes Agent::GetPrincipalAxes() const {
  PrincipalAxes principal_axes;
  // The roll axis is assumed to be aligned with the agent's velocity vector.
  principal_axes.roll = GetVelocity();
  // The pitch axis is to the agent's starboard.
  principal_axes.pitch =
      Eigen::Vector3d{principal_axes.roll(1), -principal_axes.roll(0), 0};
  // The yaw axis points upwards relative to the agent's roll-pitch plane.
  principal_axes.yaw = principal_axes.pitch.cross(principal_axes.roll);
  return principal_axes;
}

Agent::PrincipalAxes Agent::GetNormalizedPrincipalAxes() const {
  auto principal_axes = GetPrincipalAxes();
  principal_axes.roll.normalize();
  principal_axes.pitch.normalize();
  principal_axes.yaw.normalize();
  return principal_axes;
}

Eigen::Vector3d Agent::GetPosition() const {
  return Eigen::Vector3d{
      state_.position().x(),
      state_.position().y(),
      state_.position().z(),
  };
}

Eigen::Vector3d Agent::GetVelocity() const {
  return Eigen::Vector3d{
      state_.velocity().x(),
      state_.velocity().y(),
      state_.velocity().z(),
  };
}

double Agent::GetSpeed() const {
  const auto velocity = GetVelocity();
  return velocity.norm();
}

Eigen::Vector3d Agent::GetAcceleration() const {
  return Eigen::Vector3d{
      state_.acceleration().x(),
      state_.acceleration().y(),
      state_.acceleration().z(),
  };
}

Transformation Agent::GetRelativeTransformation(const Agent& agent) const {
  // TODO(titan): The transformation should be relative to the agent's roll,
  // pitch, and yaw.
  Transformation relative_transformation;

  // Find the relative position transformation.
  const auto position_transformation = GetRelativePositionTransformation(agent);
  relative_transformation.MergeFrom(position_transformation);

  // Find the relative velocity transformation.
  const auto velocity_transformation = GetRelativeVelocityTransformation(agent);
  relative_transformation.MergeFrom(velocity_transformation);

  // Find the relative acceleration transformation.
  const auto acceleration_transformation =
      GetRelativeAccelerationTransformation(agent);
  relative_transformation.MergeFrom(acceleration_transformation);

  return relative_transformation;
}

Transformation Agent::GetRelativePositionTransformation(
    const Agent& agent) const {
  Transformation position_transformation;
  const auto principal_axes = GetNormalizedPrincipalAxes();

  // Calculate the relative position of the agent.
  const auto position = GetPosition();
  const auto agent_position = agent.GetPosition();
  const auto relative_position = agent_position - position;

  // Set the Cartesian coordinates.
  position_transformation.mutable_position_cartesian()->set_x(
      relative_position(0));
  position_transformation.mutable_position_cartesian()->set_y(
      relative_position(1));
  position_transformation.mutable_position_cartesian()->set_z(
      relative_position(2));

  // Calculate the distance to the agent.
  position_transformation.mutable_position()->set_range(
      relative_position.norm());

  // Project the relative position vector onto the yaw axis.
  const auto relative_position_projection_on_yaw =
      relative_position.dot(principal_axes.yaw) * principal_axes.yaw;
  // Project the relative position vector onto the roll-pitch plane.
  const auto relative_position_projection_on_roll_pitch_plane =
      relative_position - relative_position_projection_on_yaw;

  // Determine the sign of the elevation.
  const auto elevation_sign =
      relative_position_projection_on_yaw.dot(principal_axes.yaw) >= 0 ? 1 : -1;

  // Calculate the elevation.
  position_transformation.mutable_position()->set_elevation(
      elevation_sign *
      std::atan(relative_position_projection_on_yaw.norm() /
                relative_position_projection_on_roll_pitch_plane.norm()));

  // Project the projection onto the roll axis.
  const auto relative_position_projection_on_roll =
      relative_position_projection_on_roll_pitch_plane.dot(
          principal_axes.roll) *
      principal_axes.roll;
  // Find the projection onto the pitch axis.
  const auto relative_position_projection_on_pitch =
      relative_position_projection_on_roll_pitch_plane -
      relative_position_projection_on_roll;

  if (relative_position_projection_on_pitch.norm() > 0 ||
      relative_position_projection_on_roll.norm() > 0) {
    // Determine the sign of the azimuth.
    const auto azimuth_sign =
        relative_position_projection_on_pitch.dot(principal_axes.pitch) >= 0
            ? 1
            : -1;

    // Calculate the azimuth.
    position_transformation.mutable_position()->set_azimuth(
        azimuth_sign * std::atan(relative_position_projection_on_pitch.norm() /
                                 relative_position_projection_on_roll.norm()));
  }
  return position_transformation;
}

Transformation Agent::GetRelativeVelocityTransformation(
    const Agent& agent) const {
  Transformation velocity_transformation;
  const auto principal_axes = GetNormalizedPrincipalAxes();

  // Calculate the relative position of the agent.
  const auto position = GetPosition();
  const auto agent_position = agent.GetPosition();
  const auto relative_position = agent_position - position;

  // Calculate the relative velocity of the agent.
  const auto velocity = GetVelocity();
  const auto agent_velocity = agent.GetVelocity();
  const auto relative_velocity = agent_velocity - velocity;

  // Set the Cartesian coordinates.
  velocity_transformation.mutable_velocity_cartesian()->set_x(
      relative_velocity(0));
  velocity_transformation.mutable_velocity_cartesian()->set_y(
      relative_velocity(1));
  velocity_transformation.mutable_velocity_cartesian()->set_z(
      relative_velocity(2));

  // Project the relative velocity vector onto the relative position vector.
  const auto velocity_projection_on_relative_position =
      relative_velocity.dot(relative_position) /
      std::pow(relative_position.norm(), 2) * relative_position;

  // Determine the sign of the range rate.
  const auto range_rate_sign =
      velocity_projection_on_relative_position.dot(relative_position) >= 0 ? 1
                                                                           : -1;

  // Calculate the range rate.
  velocity_transformation.mutable_velocity()->set_range(
      range_rate_sign * velocity_projection_on_relative_position.norm());

  // Project the relative velocity vector onto the sphere passing through the
  // agent.
  const auto velocity_projection_on_azimuth_elevation_sphere =
      relative_velocity - velocity_projection_on_relative_position;

  // The azimuth vector is orthogonal to the relative position vector and points
  // to the starboard of the agent along the azimuth-elevation sphere.
  auto azimuth = relative_position.cross(principal_axes.yaw);
  // The elevation vector is orthogonal to the relative position vector and
  // points upwards from the agent along the azimuth-elevation sphere.
  auto elevation = principal_axes.pitch.cross(relative_position);
  // If the relative position vector is parallel to the yaw or pitch axis, the
  // azimuth vector or the elevation vector will be undefined.
  if (azimuth.norm() == 0) {
    azimuth = relative_position.cross(elevation);
  } else if (elevation.norm() == 0) {
    elevation = azimuth.cross(relative_position);
  }

  // Project the relative velocity vector on the azimuth-elevation sphere onto
  // the azimuth vector.
  const auto velocity_projection_on_azimuth =
      velocity_projection_on_azimuth_elevation_sphere.dot(azimuth) /
      std::pow(azimuth.norm(), 2) * azimuth;

  // Determine the sign of the azimuth velocity.
  const auto azimuth_velocity_sign =
      velocity_projection_on_azimuth.dot(azimuth) >= 0 ? 1 : -1;

  // Calculate the time derivative of the azimuth.
  velocity_transformation.mutable_velocity()->set_azimuth(
      azimuth_velocity_sign * velocity_projection_on_azimuth.norm() /
      relative_position.norm());

  // Project the velocity vector on the azimuth-elevation sphere onto the
  // elevation vector.
  const auto velocity_projection_on_elevation =
      velocity_projection_on_azimuth_elevation_sphere -
      velocity_projection_on_azimuth;

  // Determine the sign of the elevation velocity.
  const auto elevation_velocity_sign =
      velocity_projection_on_elevation.dot(elevation) >= 0 ? 1 : -1;

  // Calculate the time derivative of the elevation.
  velocity_transformation.mutable_velocity()->set_elevation(
      elevation_velocity_sign * velocity_projection_on_elevation.norm() /
      relative_position.norm());
  return velocity_transformation;
}

Transformation Agent::GetRelativeAccelerationTransformation(
    const Agent& agent) const {
  Transformation acceleration_transformation;

  // Since the agent's acceleration is an input, the relative acceleration is
  // just the agent's acceleration.
  const auto agent_acceleration = agent.GetAcceleration();

  // Set the Cartesian coordinates.
  acceleration_transformation.mutable_acceleration_cartesian()->set_x(
      agent_acceleration(0));
  acceleration_transformation.mutable_acceleration_cartesian()->set_y(
      agent_acceleration(1));
  acceleration_transformation.mutable_acceleration_cartesian()->set_z(
      agent_acceleration(2));
  return acceleration_transformation;
}

Eigen::Vector3d Agent::GetGravity() const {
  return Eigen::Vector3d{
      0, 0, -constants::CalculateGravityAtAltitude(state_.position().z())};
}

double Agent::GetDynamicPressure() const {
  const auto air_density =
      constants::CalculateAirDensityAtAltitude(state_.position().z());
  const auto flow_speed = GetSpeed();
  return air_density * std::pow(flow_speed, 2) / 2;
}

void Agent::Update(const double t) {
  const auto launch_time = dynamic_config_.launch_config().launch_time();
  const auto boost_time = static_config().boost_config().boost_time();

  // Determine the flight phase.
  if (t >= t_creation_ + launch_time) {
    flight_phase_ = FlightPhase::BOOST;
  }
  if (t >= t_creation_ + launch_time + boost_time) {
    flight_phase_ = FlightPhase::MIDCOURSE;
  }
  // TODO(titan): Determine when to enter the terminal phase.

  switch (flight_phase_) {
    case FlightPhase::INITIALIZED: {
      return;
    }
    case FlightPhase::READY: {
      UpdateReady(t);
      break;
    }
    case FlightPhase::BOOST: {
      UpdateBoost(t);
      break;
    }
    case FlightPhase::MIDCOURSE:
    case FlightPhase::TERMINAL: {
      UpdateMidCourse(t);
      break;
    }
    default: {
      throw std::runtime_error(
          absl::StrFormat("Invalid flight phase: %d.", flight_phase_));
    }
  }
}

void Agent::Step(const double t_start, const double t_step) {
  // TODO(titan): Evolve the agent's roll, pitch, and yaw.
  using Vector6d = Eigen::Matrix<double, 6, 1>;

  // Update the latest state in the history.
  state_history_.back().t = t_start;
  state_history_.back().state = state_;

  // Check if the step time is zero.
  if (t_step == 0) {
    return;
  }

  const auto position = GetPosition();
  const auto velocity = GetVelocity();

  // The state vector consists of the position and the velocity vectors.
  Vector6d x{
      position(0), position(1), position(2),
      velocity(0), velocity(1), velocity(2),
  };

  // Define the state-space equations.
  const auto kinematics = [&](const Vector6d& x, Vector6d& x_dot,
                              const double t) {
    const auto position_z = x(2);
    const auto velocity_x = x(3);
    const auto velocity_y = x(4);
    const auto velocity_z = x(5);

    // Check if the agent has hit the ground.
    if (position_z < 0) {
      x_dot = Vector6d::Zero();
    } else {
      x_dot = Vector6d{
          // dx/dt = vx
          velocity_x,
          // dy/dt = vy
          velocity_y,
          // dz/dt = vz
          velocity_z,
          // dvx/dt = ax
          state_.acceleration().x(),
          // dvy/dt = ay
          state_.acceleration().y(),
          // dvz/dt = az
          state_.acceleration().z(),
      };
    }
  };

  // Integrate the state-space equations.
  const auto t_end = t_start + t_step;
  boost::numeric::odeint::integrate(kinematics, x, /*start_time=*/t_start,
                                    /*end_time=*/t_end, /*dt=*/t_step / 10);

  // Set the new state.
  const auto position_x = x(0);
  const auto position_y = x(1);
  const auto position_z = x(2);
  const auto velocity_x = x(3);
  const auto velocity_y = x(4);
  const auto velocity_z = x(5);
  state_.mutable_position()->set_x(position_x);
  state_.mutable_position()->set_y(position_y);
  state_.mutable_position()->set_z(position_z);
  state_.mutable_velocity()->set_x(velocity_x);
  state_.mutable_velocity()->set_y(velocity_y);
  state_.mutable_velocity()->set_z(velocity_z);

  // Add the new state to the history.
  state_history_.Add(state::StateHistory::Record(t_end, hit_, state_));
  state_update_time_ = t_end;
}

}  // namespace swarm::agent

File: simulation/swarm/py/agent.py
"""The agent class is an interface for an interceptor or a threat."""

from abc import ABC, abstractmethod
from collections import namedtuple
from typing import Self

import numpy as np
import scipy.integrate
from simulation.swarm.proto.agent_pb2 import AgentConfig, FlightPhase
from simulation.swarm.proto.dynamic_config_pb2 import DynamicConfig
from simulation.swarm.proto.plotting_config_pb2 import PlottingConfig
from simulation.swarm.proto.state_pb2 import State
from simulation.swarm.proto.static_config_pb2 import StaticConfig

from simulation.swarm.utils.py import constants


class Agent(ABC):
    """Agent.

    Attributes:
        t_creation: The agent's creation time in s.
        state: The current state.
        state_update_time: The time of the last state update in s.
        flight_phase: The flight phase of the agent.
        static_config: The static configuration of the agent.
        dynamic_config: The dynamic configuration of the agent.
        plotting_config: The plotting configuration of the agent.
        submunitions_config: The submunitions configuration of the agent.
        history: A list of 3-tuples consisting of a timestamp, the hit boolean,
          and the state.
        hit: A boolean indicating whether the agent has hit or been hit.
    """

    # History record named tuple type.
    HistoryRecord = namedtuple("HistoryRecord", ["t", "hit", "state"])

    def __init__(
        self,
        config: AgentConfig = None,
        ready: bool = True,
        t_creation: float = 0,
        *,
        initial_state: State = None,
        dynamic_config: DynamicConfig = None,
        plotting_config: PlottingConfig = None,
        submunitions_config: AgentConfig.SubmunitionsConfig = None,
    ) -> None:
        self.t_creation = t_creation

        # Set the initial state.
        self.state = State()
        if initial_state is not None:
            self.state.CopyFrom(initial_state)
        else:
            self.state.CopyFrom(config.initial_state)
        self.state_update_time = 0
        # In the initialized flight phase, the agent idles, but in the ready
        # flight phase, the agent is subject to physical forces.
        self.flight_phase = (FlightPhase.READY
                             if ready else FlightPhase.INITIALIZED)

        # Set the dynamic configuration.
        self.dynamic_config = DynamicConfig()
        if dynamic_config is not None:
            self.dynamic_config.CopyFrom(dynamic_config)
        elif config is not None:
            self.dynamic_config.CopyFrom(config.dynamic_config)

        # Set the plotting configuration.
        self.plotting_config = PlottingConfig()
        if plotting_config is not None:
            self.plotting_config.CopyFrom(plotting_config)
        elif config is not None:
            self.plotting_config.CopyFrom(config.plotting_config)

        # Set the submunitions configuration.
        self.submunitions_config = AgentConfig.SubmunitionsConfig()
        if submunitions_config is not None:
            self.submunitions_config.CopyFrom(submunitions_config)
        elif config is not None:
            self.submunitions_config.CopyFrom(config.submunitions_config)

        self.hit = False
        self.history = [
            Agent.HistoryRecord(t=self.t_creation, hit=self.hit, state=State())
        ]
        self.history[-1].state.CopyFrom(self.state)

    @property
    @abstractmethod
    def static_config(self) -> StaticConfig:
        """Returns the static configuration of the agent."""

    def has_launched(self) -> bool:
        """Returns whether the agent has launched."""
        return (self.flight_phase != FlightPhase.INITIALIZED and
                self.flight_phase != FlightPhase.READY)

    def has_terminated(self) -> bool:
        """Returns whether the agent's flight has terminated."""
        return self.flight_phase == FlightPhase.TERMINATED

    def mark_as_hit(self) -> None:
        """Sets the agent to have hit the target or have been hit."""
        self.hit = True
        # Update the latest hit boolean in the history of states.
        self.history[-1] = self.history[-1]._replace(hit=True)
        self.flight_phase = FlightPhase.TERMINATED

    def set_state(self, state: State) -> None:
        """Sets the state of the agent."""
        self.state.CopyFrom(state)
        # Update the latest state in the history of states.
        self.history[-1].state.CopyFrom(state)

    def get_principal_axes(self) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
        """Returns the principal axes of the agent.

        The principal axis directions have not been normalized.
        The roll axis is assumed to be aligned with the agent's velocity vector,
        and the pitch axis points to the starboard of the agent.
        TODO(titan): The roll axis does not necessarily have to be aligned with
        the agent's velocity vector.

        Returns:
            A 3-tuple consisting of the roll, pitch, and yaw axes.
        """
        # The roll axis is assumed to be aligned with the agent's velocity
        # vector.
        roll = self.get_velocity()
        # The pitch axis is to the agent's starboard.
        pitch = np.array([roll[1], -roll[0], 0])
        # The yaw axis points upwards relative to the agent's roll-pitch plane.
        yaw = np.cross(pitch, roll)
        return roll, pitch, yaw

    def get_normalized_principal_axes(
            self) -> tuple[np.ndarray, np.ndarray, np.ndarray]:
        """Returns the normalized_principal axes of the agent.

        Returns:
            A 3-tuple consisting of the normalized roll, pitch, and yaw axes.
        """
        roll, pitch, yaw = self.get_principal_axes()
        normalized_roll = roll / np.linalg.norm(roll)
        normalized_pitch = pitch / np.linalg.norm(pitch)
        normalized_yaw = yaw / np.linalg.norm(yaw)
        return normalized_roll, normalized_pitch, normalized_yaw

    def get_position(self) -> np.ndarray:
        """Returns the position vector of the agent."""
        position = np.array([
            self.state.position.x,
            self.state.position.y,
            self.state.position.z,
        ])
        return position

    def get_velocity(self) -> np.ndarray:
        """Returns the velocity vector of the agent."""
        velocity = np.array([
            self.state.velocity.x,
            self.state.velocity.y,
            self.state.velocity.z,
        ])
        return velocity

    def get_speed(self) -> float:
        """Returns the speed of the agent."""
        velocity = self.get_velocity()
        speed = np.linalg.norm(velocity)
        return speed

    def get_gravity(self) -> np.ndarray:
        """Returns the gravity acceleration vector."""
        gravity = np.array([
            0,
            0,
            -constants.gravity_at_altitude(self.state.position.z),
        ])
        return gravity

    def get_dynamic_pressure(self) -> float:
        """Calculates the dynamic air pressure around the agent."""
        air_density = constants.air_density_at_altitude(self.state.position.z)
        flow_speed = self.get_speed()
        dynamic_pressure = 1 / 2 * air_density * flow_speed**2
        return dynamic_pressure

    def update(self, t: float) -> None:
        """Updates the agent's state according to the environment.

        Args:
            t: Time in seconds.
        """
        launch_time = self.dynamic_config.launch_config.launch_time
        boost_time = self.static_config.boost_config.boost_time

        # Determine the flight phase.
        if t >= self.t_creation + launch_time:
            self.flight_phase = FlightPhase.BOOST
        if t >= self.t_creation + launch_time + boost_time:
            self.flight_phase = FlightPhase.MIDCOURSE
        # TODO(titan): Determine when to enter the terminal phase.

        match self.flight_phase:
            case FlightPhase.INITIALIZED:
                return
            case FlightPhase.READY:
                self._update_ready(t)
            case FlightPhase.BOOST:
                self._update_boost(t)
            case FlightPhase.MIDCOURSE | FlightPhase.TERMINAL:
                self._update(t)
            case FlightPhase.TERMINATED:
                return
            case _:
                raise ValueError(f"Invalid flight phase: {self.flight_phase}.")

    def step(self, t_start: float, t_step: float) -> None:
        """Steps forward the simulation by simulating the dynamics of the
        agent.

        This function mutates the agent's state variable.
        TODO(titan): Evolve the agent's roll, pitch, and yaw.

        Args:
            t_start: Start time in seconds.
            t_step: Step time in seconds.
        """
        # Update the latest state in the history of states.
        self.history[-1] = self.history[-1]._replace(t=t_start)
        self.history[-1].state.CopyFrom(self.state)

        # Check if the step time is zero.
        if t_step == 0:
            return

        position = self.get_position()
        velocity = self.get_velocity()
        initial_state = np.concatenate((position, velocity))

        def kinematics(t: float, state: np.ndarray) -> np.ndarray:
            """Defines the kinematic equations of the agent.

            Args:
                t: Time in seconds.
                state: Current state vector.

            Returns:
                The time derivative of the state vector at the given time.
            """
            (
                position_x,
                position_y,
                position_z,
                velocity_x,
                velocity_y,
                velocity_z,
            ) = state
            if position_z < 0:
                dx = np.zeros(state.shape)
            else:
                dx = np.array([
                    # dx/dt = vx
                    velocity_x,
                    # dy/dt = vy
                    velocity_y,
                    # dz/dt = vz
                    velocity_z,
                    # dvx/dt = ax
                    self.state.acceleration.x,
                    # dvy/dt = ay
                    self.state.acceleration.y,
                    # dvz/dt = az
                    self.state.acceleration.z,
                ])
            return dx

        solution = scipy.integrate.solve_ivp(
            fun=kinematics,
            t_span=[0, t_step],
            y0=initial_state,
            t_eval=[t_step],
        )
        (
            position_x,
            position_y,
            position_z,
            velocity_x,
            velocity_y,
            velocity_z,
        ) = np.squeeze(solution.y)
        self.state.position.x = position_x
        self.state.position.y = position_y
        self.state.position.z = position_z
        self.state.velocity.x = velocity_x
        self.state.velocity.y = velocity_y
        self.state.velocity.z = velocity_z

        # Add the new state to the history of states.
        t = t_start + t_step
        self.history.append(
            Agent.HistoryRecord(t=t, hit=self.hit, state=State()))
        self.history[-1].state.CopyFrom(self.state)
        self.state_update_time = t

    def spawn(self, t: float) -> list[Self]:
        """Spawns new agents.

        Args:
            t: Time in seconds.

        Returns:
            A list of newly spawned agents.
        """
        return []

    def _update_ready(self, t: float) -> None:
        """Updates the agent's state in the ready flight phase.

        Args:
            t: Time in seconds.
        """
        # By default, idle in the ready flight phase.
        return

    def _update_boost(self, t: float) -> None:
        """Updates the agent's state in the boost flight phase.

        Args:
            t: Time in seconds.
        """
        # By default, idle in the boost flight phase.
        return

    def _update(self, t: float) -> None:
        """Updates the agent's state in the midcourse and terminal flight
        phase.

        Args:
            t: Time in seconds.
        """
        # By default, idle in the midcourse and terminal flight phase.
        return
